core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheManager.java

```diff
@@ -11,74 +11,251 @@
 
 package alluxio.client.file.cache;
 
+import alluxio.client.file.FileSystemContext;
+import alluxio.collections.Pair;
+import alluxio.conf.PropertyKey;
+import alluxio.resource.LockResource;
+import alluxio.util.io.BufferUtils;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.nio.channels.Channels;
 import java.nio.channels.ReadableByteChannel;
 import java.nio.channels.WritableByteChannel;
+import java.util.Arrays;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 
+import javax.annotation.concurrent.GuardedBy;
 import javax.annotation.concurrent.ThreadSafe;
 
 /**
- * A class to manage cached pages. This class will
- * 1. Ensure thread-safety
- * 2. Bookkeep Cache Replacement Alg
+ * A class to manage cached pages. This class coordinates different components to respond for
+ * thread-safety and operate cache replacement policies.
  *
+ * Lock hierarchy in this class: All operations must follow this order to operate on pages:
+ * <ul>
+ * <li>1. Acquire page lock</li>
+ * <li>2. Acquire metastore lock mMetaLock</li>
+ * <li>3. Release metastore lock mMetaLock</li>
+ * <li>4. Release page lock</li>
+ * </ul>
  */
 @ThreadSafe
 public class LocalCacheManager {
   private static final Logger LOG = LoggerFactory.getLogger(LocalCacheManager.class);
+  /** Number of page locks to strip. */
+  private static int PAGE_LOCK_SIZE = 256;
```

Comment:
do we ever need to make this configurable?

Reply:
Doesn't hurt to add a property.

Reply:
this is a very internal config and I don't expect the users to tune this value.
Given we have already too many configuration keys, I would leave this a constant large enough, e.g., a few thousands.